cellmlmanip/model.py

```diff
@@ -73,52 +73,67 @@ def __str__(self):
+
+    The main parts of a Model are 1. a list of sympy equation objects; 2. a collection of named units; and 3. and RDF
+    graph that stores further meta data about the Model.
 
-    :param name: the name of the model e.g. from <model name="">
-    :param cmeta_id: the cmeta_id of the model e.g. from <model cmeta_id="">
+    Equations are stored as ``Sympy.Eq`` objects, but with the caveat that all variables and numbers must be specified
+    using the ``Sympy.Dummy`` objects returned by :meth:`add_variable()` and :meth:`add_number()`.
+
+    :param name: the name of the model e.g. from ``<model name="">``.
+    :param cmeta_id: An optional ``cmeta_id``, e.g. from ``<model cmeta_id="">``.
     """
-    def __init__(self, name, cmeta_id):
+    def __init__(self, name, cmeta_id=None):
 
         self.name = name
         self.cmeta_id = cmeta_id
         self.rdf_identity = rdflib.URIRef('#' + cmeta_id) if cmeta_id else None
+
+        # A list of sympy.Eq equation objects
+        self.equations = []
+
+        # A pint UnitStore, mapping unit names to unit objects
         self.units = UnitStore(model=self)
+
+        # An RDF graph containing further meta data
         self.rdf = rdflib.Graph()
-        self.graph = None   # An nx.DiGraph
 
         # Maps sympy.Dummy objects to MetaDummy objects
         self.dummy_metadata = OrderedDict()
 
         # Maps string names to sympy.Dummy objects
         self.name_to_symbol = dict()
 
-        # A list of sympy.Eq objects.
-        self.equations = []
+        # A cached nx.DiGraph of this model's equations, created by get_equation_graph()
+        self.graph = None
 
-    def add_unit(self, units_name, unit_attributes=None, base_units=False):
+    def add_unit(self, name, attributes=None, base_units=False):
         """
-        Adds information about <units> in <model>.
+        Adds a unit of measurement to this model, with a given ``name`` and list of ``attributes``.
 
         :param units_name: A string name
         :param unit_attributes: An optional list of dictionaries containing unit attributes. See
             :meth:`UnitStore.add_custom_unit()`.
         :base_units: Set to ``True`` to define a new base unit.
         """
-        assert not (unit_attributes and base_units), 'Cannot define base unit with unit attributes'
         if base_units:
-            self.units.add_base_unit(units_name)
+            if attributes:
+                raise ValueError('Base units can not be defined with unit attributes.')
+            self.units.add_base_unit(name)
         else:
-            self.units.add_custom_unit(units_name, unit_attributes)
+            self.units.add_custom_unit(name, attributes)
 
     def add_equation(self, equation):
         """
         Adds an equation to this model.
 
         :param equation: A ``sympy.Eq`` object.
         """
-        assert isinstance(equation, sympy.Eq), 'Equation expression must be equality'
+        if not isinstance(equation, sympy.Eq):
```

Comment:
I think for this kind of sanity check asserts are OK. You could get user error, but it's not a recoverable error, so assert feels more natural than exception. The same applies to the ones below. The one above I'd be more OK with changing to an exception.

We'll also need lots more silly tests to maintain coverage if we change!

Reply:
Really? There's at least two built-in exception types for dealing with input specifically https://docs.python.org/3/library/exceptions.html#TypeError

> We'll also need lots more silly tests to maintain coverage if we change!

But that's a good thing surely? Asserts just hide the fact that this part of the program is untested. If you really wanted to we could always add coverage exclude rules for any `raise` statement

Reply:
True, we could raise `TypeError` rather than asserting. I don't think there's a strong case either way, and `assert` is lower maintenance, which I why I favour it here; at least until the API settles down and we have an external user!